extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/DevServicesKubernetesProcessor.java

```diff
@@ -0,0 +1,415 @@
+            }
+            if (config.serviceName != null) {
+                container.withLabel(DevServicesKubernetesProcessor.DEV_SERVICE_LABEL, config.serviceName);
+            }
+            timeout.ifPresent(container::withStartupTimeout);
+
+            container.start();
+
+            KubeConfig kubeConfig = KubeConfigUtils.parseKubeConfig(container.getKubeconfig());
+
+            return new RunningDevService(Feature.KUBERNETES_CLIENT.getName(), container.getContainerId(),
+                    new ContainerShutdownCloseable(container, Feature.KUBERNETES_CLIENT.getName()),
+                    getKubernetesClientConfigFromKubeConfig(kubeConfig));
+        };
+
+        return maybeContainerAddress
+                .map(containerAddress -> new RunningDevService(Feature.KUBERNETES_CLIENT.getName(),
+                        containerAddress.getId(),
+                        null,
+                        resolveConfigurationFromRunningContainer(containerAddress)))
+                .orElseGet(defaultKubernetesClusterSupplier);
+    }
+
+    <T extends KubernetesVersionEnum<T>> T findOrElseThrow(final String version, final Class<T> versions) {
+        final String versionWithPrefix = !version.startsWith("v") ? "v" + version : version;
+        return Stream.of(versions.getEnumConstants())
+                .filter(v -> v.descriptor().getKubernetesVersion().equalsIgnoreCase(versionWithPrefix))
+                .findFirst()
+                .orElseThrow();
+    }
+
+    private Map<String, String> getKubernetesClientConfigFromKubeConfig(KubeConfig kubeConfig) {
+
+        ClusterSpec cluster = kubeConfig.getClusters().get(0).getCluster();
+        UserSpec user = kubeConfig.getUsers().get(0).getUser();
+        return Map.of(
+                KUBERNETES_CLIENT_MASTER_URL, cluster.getServer(),
+                "quarkus.kubernetes-client.ca-cert-data",
+                cluster.getCertificateAuthorityData(),
+                "quarkus.kubernetes-client.client-cert-data",
+                user.getClientCertificateData(),
+                "quarkus.kubernetes-client.client-key-data", user.getClientKeyData(),
+                "quarkus.kubernetes-client.client-key-algo", Config.getKeyAlgorithm(null, user.getClientKeyData()),
+                "quarkus.kubernetes-client.namespace", "default");
+    }
+
+    private Map<String, String> resolveConfigurationFromRunningContainer(ContainerAddress containerAddress) {
+        var dockerClient = DockerClientFactory.lazyClient();
+        var container = new RunningContainer(dockerClient, containerAddress);
+
+        return container.getKubeconfig();
+    }
+
+    private KubernetesDevServiceCfg getConfiguration(KubernetesClientBuildConfig cfg) {
+        KubernetesDevServicesBuildTimeConfig devServicesConfig = cfg.devservices;
+        return new KubernetesDevServiceCfg(devServicesConfig);
+    }
+
+    private static final class KubernetesDevServiceCfg {
```

Comment:
I never implemented a dev service, but why do we need this class instead of using directly KubernetesClientBuildConfig?

Reply:
I reuse the logic from elsewhere. The difference is that the class must implements equals to be able to detect changes.